nova/src/ccs/mod.rs

```diff
@@ -0,0 +1,532 @@
+use ark_ec::{AdditiveGroup, CurveGroup};
+use ark_ff::Field;
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use ark_spartan::polycommitments::PolyCommitmentScheme;
+
+use ark_std::Zero;
+use std::ops::Neg;
+
+#[cfg(feature = "parallel")]
+use rayon::iter::{
+    IndexedParallelIterator, IntoParallelIterator, IntoParallelRefIterator,
+    IntoParallelRefMutIterator, ParallelIterator,
+};
+
+use super::r1cs::R1CSShape;
+pub use super::sparse::{MatrixRef, SparseMatrix};
+use mle::vec_to_mle;
+
+pub mod mle;
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum Error {
+    InvalidWitnessLength,
+    InvalidInputLength,
+    InvalidEvaluationPoint,
+    InvalidTargets,
+    NotSatisfied,
+}
+
+/// A type that holds the shape of the CCS matrices
+#[derive(Debug, Clone, PartialEq, Eq, CanonicalSerialize, CanonicalDeserialize)]
+pub struct CCSShape<G: CurveGroup> {
+    /// `m` in the CCS/HyperNova papers.
+    pub num_constraints: usize,
+    /// Witness length.
+    ///
+    /// `m - l - 1` in the CCS/HyperNova papers.
+    pub num_vars: usize,
+    /// Length of the public input `X`. It is expected to have a leading
+    /// `ScalarField::ONE` element, thus this field must be non-zero.
+    ///
+    /// `l + 1`, w.r.t. the CCS/HyperNova papers.
+    pub num_io: usize,
+    /// Number of matrices.
+    ///
+    /// `t` in the CCS/HyperNova papers.
+    pub num_matrices: usize,
+    /// Number of multisets.
+    ///
+    /// `q` in the CCS/HyperNova papers.
+    pub num_multisets: usize,
+    /// Max cardinality of the multisets.
+    ///
+    /// `d` in the CCS/HyperNova papers.
+    pub max_cardinality: usize,
+    /// Set of constraint matrices.
+    pub Ms: Vec<SparseMatrix<G::ScalarField>>,
+    /// Multisets of selector indices, each paired with a constant multiplier.
+    pub cSs: Vec<(G::ScalarField, Vec<usize>)>,
+}
+
+impl<G: CurveGroup> CCSShape<G> {
+    /// Checks if the CCS instance together with the witness `W` satisfies the CCS constraints determined by `shape`.
+    pub fn is_satisfied<C: PolyCommitmentScheme<G>>(
```

Comment:
When we implement CycleFold, we'll also need to consider CCS instances over the secondary field. The commitments for these will just be Pedersen commitments, not coming from a polynomial commitment scheme. (KZG-style commitments won't exist for the secondary curve, because it doesn't support pairings). On the other hand, it is indeed important - for compression - that the commitment scheme for the primary curve does come from a PCS. In the nova repo, we define everything in terms of vector commitment schemes, and then to create the "compressible" version, we coerce the PCS for the primary field into a vector commitment scheme. Another approach would be to just define separate types, e.g. `CCSInstanceSecondary<G: CurveGroup, C: CommitmentScheme<G>>` - it's a bit redundant, but maybe cleaner.